src/meta/src/stream/stream_graph.rs

```diff
@@ -1140,53 +1244,177 @@ impl CompleteStreamFragmentGraph {
             .collect();
 
         Ok(Self {
-            graph,
+            building_graph: graph,
             existing_fragments,
-            downstreams,
-            upstreams,
+            extra_downstreams,
+            extra_upstreams,
         })
     }
 
-    /// Returns the dispatcher types of all edges in the complete graph.
-    fn dispatch_edges(
+    /// Returns **all** fragment IDs in the complete graph, including the ones that are not in the
+    /// building graph.
+    fn all_fragment_ids(&self) -> impl Iterator<Item = GlobalFragmentId> + '_ {
+        self.building_graph
+            .fragments
+            .keys()
+            .chain(self.existing_fragments.keys())
+            .copied()
+    }
+
+    /// Returns an iterator of **all** edges in the complete graph, including the external edges.
+    fn all_edges(
         &self,
-    ) -> impl Iterator<Item = (GlobalFragmentId, GlobalFragmentId, DispatcherType)> + '_ {
-        self.graph
-            .edges()
-            .map(|(from, to, edge)| {
-                let dt = edge.get_dispatch_strategy().unwrap().get_type().unwrap();
-                (from, to, dt)
-            })
-            .chain(
-                self.downstreams
-                    .iter()
-                    .flat_map(|(&from, tos)| tos.iter().map(move |(&to, &dt)| (from, to, dt))),
-            )
+    ) -> impl Iterator<Item = (GlobalFragmentId, GlobalFragmentId, &StreamFragmentEdge)> + '_ {
+        self.building_graph
+            .downstreams
+            .iter()
+            .chain(self.extra_downstreams.iter())
+            .flat_map(|(&from, tos)| tos.iter().map(move |(&to, edge)| (from, to, edge)))
     }
 
     /// Returns the distribution of the existing fragments.
     fn existing_distribution(&self) -> HashMap<GlobalFragmentId, Distribution> {
         self.existing_fragments
             .iter()
-            .map(|(&id, f)| {
-                let dist = match f.get_distribution_type().unwrap() {
-                    FragmentDistributionType::Unspecified => unreachable!(),
-                    FragmentDistributionType::Single => Distribution::Singleton,
-                    FragmentDistributionType::Hash => {
-                        let mapping =
-                            ParallelUnitMapping::from_protobuf(f.get_vnode_mapping().unwrap());
-                        Distribution::Hash(mapping)
-                    }
-                };
-                (id, dist)
-            })
+            .map(|(&id, f)| (id, Distribution::from_fragment(f)))
             .collect()
     }
 
-    /// Consumes this complete graph and returns the inner graph.
-    // TODO: remove this after scheduler refactoring.
-    pub fn into_inner(self) -> StreamFragmentGraph {
-        self.graph
+    /// Generate topological order of **all** fragments in this graph, including the ones that are
```

Comment:
The interfaces below used to be the `FragmentGraph`'s. Now it's modified to operate on the `CompleteGraph`, where the existing fragments will be considered.